awscrt/s3.py

```diff
@@ -0,0 +1,204 @@
+
+                *   `**kwargs` (dict): Forward-compatibility kwargs.
+
+            on_body: Optional callback invoked 0+ times as the response body received from S3 server.
+                The function should take the following arguments and return nothing:
+
+                *   `chunk` (buffer): Response body data (not necessarily
+                    a whole "chunk" of chunked encoding).
+
+                *   `offset` (int): The offset of the chunk started in the whole body.
+
+                *   `**kwargs` (dict): Forward-compatibility kwargs.
+
+            on_done: Optional callback invoked when the meta_request has finished the job.
+                The function should take the following arguments and return nothing:
+
+                *   `error_code` (int): The error code from CRT, indicates the job has finished successfully
+                    or failed for specific reason
+
+                *   `**kwargs` (dict): Forward-compatibility kwargs.
+
+        Returns:
+            S3Request
+        """
+        return S3Request(
+            client=self,
+            request=request,
+            type=type,
+            on_headers=on_headers,
+            on_body=on_body,
+            on_done=on_done,
+            **kwargs)
+
+
+class S3Request(NativeResource):
+    """S3 request
+
+    """
+    __slots__ = ('_on_headers_cb', '_on_body_cb', '_on_done_cb', '_finished_future', '_kwargs', 'shutdown_event')
+
+    def __init__(self, *, client, request, type, on_headers=None, on_body=None, on_done=None, **kwargs):
+        assert isinstance(client, S3Client)
+        assert isinstance(request, HttpRequest)
+        assert callable(on_headers) or on_headers is None
+        assert callable(on_body) or on_body is None
+        assert callable(on_done) or on_done is None
+
+        super().__init__()
+
+        # the native s3-request will keep the request alive until the s3-request finishes
+        self._on_headers_cb = on_headers
+        self._on_body_cb = on_body
+        self._on_done_cb = on_done
+        self._kwargs = kwargs
+
+        self._finished_future = Future()
+
+        shutdown_event = threading.Event()
+
+        def on_shutdown():
+            shutdown_event.set()
+
+        self.shutdown_event = shutdown_event
+
+        self._binding = _awscrt.s3_client_make_meta_request(
+            self, client, request, type, self._on_headers, self._on_body, self._on_finish, on_shutdown)
+
+    def _on_headers(self, status_code, headers):
+        if self._on_headers_cb:
+            self._on_headers_cb(status_code=status_code, headers=headers, **self._kwargs)
+
+    def _on_body(self, chunk, offset):
+        if self._on_body_cb:
+            self._on_body_cb(chunk=chunk, offset=offset, kwargs=self._kwargs)
+
+    def _on_finish(self, error_code):
+        if error_code:
+            self.finished_future.set_exception(awscrt.exceptions.from_code(error_code))
+        else:
+            self.finished_future.set_result("finish")
```

Comment:
trivial: In HTTP, we set a similar future's result to be the `status_code`
could do the same here

Reply:
feels like we can just set no result since the invalid response status code will result in a failure from native client.